bootstore/src/schemes/v0/peer.rs

```diff
@@ -0,0 +1,810 @@
+                    Ok(Some(api_output)) => {
+                        self.handle_api_output(api_output).await
+                    }
+                    Err(err) => self.handle_api_error(err).await,
+                }
+            }
+            ConnToMainMsgInner::FailedServerHandshake {
+                addr: SocketAddrV6,
+            } => {}
+        }
+    }
+
+    async fn manage_connections(&mut self, peers: BTreeSet<SocketAddrV6>) {
+        if peers == self.peers {
+            return;
+        }
+
+        let peers_to_remove: BTreeSet<_> =
+            self.peers.difference(&peers).cloned().collect();
+        let new_peers: BTreeSet<_> =
+            peers.difference(&self.peers).cloned().collect();
+
+        self.peers = peers;
+
+        // Start a new client for each peer that has an addr < self.config.addr
+        for addr in new_peers {
+            if addr < self.config.addr {
+                self.handle_unique_id_counter += 1;
+                let handle = spawn_connection_initiator_task(
+                    self.handle_unique_id_counter,
+                    self.config.id.clone(),
+                    self.config.addr.clone(),
+                    addr.clone(),
+                    &self.log,
+                    self.conn_tx.clone(),
+                )
+                .await;
+                self.initiating_connections.insert(addr, handle);
+            }
+        }
+
+        // Remove each peer that we no longer need a connection to
+        for addr in peers_to_remove {
+            self.remove_peer(addr).await;
+        }
+    }
+
+    async fn remove_accepted_connection(&mut self, addr: &SocketAddrV6) {
+        if let Some(handle) = self.accepted_connections.remove(&addr) {
+            // The connection has not yet completed its handshake
+            let _ = handle.tx.send(MainToConnMsg::Close).await;
+        }
+    }
+
+    async fn remove_peer(&mut self, addr: SocketAddrV6) {
+        if let Some(handle) = self.initiating_connections.remove(&addr) {
+            // The connection has not yet completed its handshake
+            let _ = handle.tx.send(MainToConnMsg::Close).await;
+        } else {
+            // Do we have an established connection?
+            if let Some((id, handle)) = self
+                .established_connections
+                .iter()
+                .find(|(_, handle)| handle.addr == addr)
+            {
+                let _ = handle.tx.send(MainToConnMsg::Close).await;
+                // probably a better way to avoid borrowck issues
+                let id = id.clone();
+                self.established_connections.remove(&id);
+            }
+        }
+    }
+
+    async fn remove_established_connection(&mut self, peer_id: &Baseboard) {
+        if let Some(handle) = self.established_connections.remove(peer_id) {
+            // Gracefully stop the task
+            let _ = handle.tx.send(MainToConnMsg::Close).await;
+        }
+    }
```

Comment:
Worth adding a warning here if a connection isn't found in the `established_connections` map.

Reply:
In the common case the connections shouldn't exist, as it means a connecting client connected twice. The only way they would exist is if there is a race, and the remote client detects shutdown first. In that case, I think no warning is preferred.